### DISCUSSION TITLE
Visualising Pallas Load/Stores (even with BlockSpecs)

### DISCUSSION THREAD

**Discussion by oliverdutton** [jax-ml/jax#22300] [Show and tell]
[Pallas_Visualisation](https://github.com/oliverdutton/pallas_visualisation/tree/main)

A tool to visualise the load/store ops in Pallas functions. Similar to [triton-viz](https://github.com/Deep-Learning-Profiling-Tools/triton-viz) which the visualisation 

You can use any pallas machinery for writing your kernel, e.g. BlockSpecs or the grid, and visualise it.

[![Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/oliverdutton/pallas_visualisation/blob/main/examples/demo.ipynb)

You can view indexing for a particular index on the grid or an overview of the whole grid, such as this add op split over four blocks.

The tracing is done by adding MemRefs which track the indexing, on which all load/store ops are duplicated to operate on and then logged in an external callback by editing the [jaxpr](https://jax.readthedocs.io/en/latest/jaxpr.html).

# As a followup
Would anyone be interested in:
1.  a Pallas version of [Triton-Puzzles](https://github.com/srush/Triton-Puzzles) but with Pallas explicitly in mind from the start.     
or   
2. BlockSpecs that support masking (so a boundary mask is added to any loads/stores and the 'other' kwarg is set on a per-MemRef rather than per-load basis) and implicit tiling (for the subset of cases where we are just working with blocks which partition the arrays, to reduce the amount of code required)
